File: 0078-subsets/0078-subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        result=[]
        subset=[]
        def backtrack(index):
            if index == len(nums):
                result.append(subset.copy())
                return

            
            subset.append(nums[index])
            backtrack(index+1)
            subset.pop()
            backtrack(index+1)
        backtrack(0)
        return result

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    vector<vector<int>> ans;

    void recursion(int idx, vector<int>& nums, vector<int>& res) {
        // store current subset
        ans.push_back(res);

        for(int i = idx; i < nums.size(); i++) {
            res.push_back(nums[i]);          // choose
            recursion(i + 1, nums, res);    // explore
            res.pop_back();                 // backtrack
        }
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> res;
        recursion(0, nums, res);
        return ans;
    }
};